File: 프로그래머스/lv3/42895. N으로 표현/N으로 표현.cpp
#include <string>
#include <vector>
#include <unordered_set>

using namespace std;

vector<unordered_set<int>> dp(9);

int solution(int N, int number) {   
    if(N == number) return 1;
    dp[1].insert(N);
    
    for(int step = 2; step <= 8; step++) {
        int tmp = N;
        for(int i = 2; i <= step; i++) {
            tmp = tmp * 10 + N;
        }
        dp[step].insert(tmp);
        
        for(int i = 1; i < step; i++) {
            for(int a: dp[i]) {
                for(int b: dp[step - i]) {
                    dp[step].insert(a + b);
                    dp[step].insert(a * b);
                    if(a - b > 0) dp[step].insert(a - b);
                    if(a / b > 0) dp[step].insert(a / b);
                }
            }
        }
        
        if(dp[step].find(number) != dp[step].end()) {
            return step;
        }
    }
    
    return -1;
}

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.py
def solution(N, number):
    dp = [set() for _ in range(9)]
    
    for step in range(1, 9):
        base = 0
        for _ in range(step):
            base = base * 10 + N
            
        dp[step].add(base)
        
        for i in range(1, step):
            for n in dp[i]:
                for m in dp[step - i]:
                    dp[step].add(n + m)
                    dp[step].add(n * m)
                    if n - m > 0:
                        dp[step].add(n - m)
                    if n // m > 0:
                        dp[step].add(n // m)
                        
        if number in dp[step]:
            return step
    
    return -1